### PR TITLE
Change opencv to tinyscaler

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
Tinyscaler is a new module specially built to provide an efficient and small module that solely resizes / scales NumPy ndarrays

This is backward compatibility breaking PR as all resizing implementations are unique so tinyscaler does not produce the same ndarrays as opencv however opencv is an annoying module that we have wanted to change for a while.

### CODE REVIEW

**Comment by 222464** [comment]
Here are some test results with rl-baselines3-zoo for dqn and ppo on some Atari games. The TinyScaler area filtering variants 1 and 2 are not as "correct feeling" as variant 3, which is the one on GitHub, but they perform better on some Atari games. To a human they all look about the same when compared to OpenCV, however especially for Breakout dqn seems to have trouble when using TinyScaler. I believe this is because the ball is really small in Breakout (as opposed to Pong), making it more susceptible to sampling pattern differences. Interestingly ppo seems less bothered by it.

### OpenCV area filtering results (baseline)

|algo|        env_id         |    mean_reward     |     std_reward      |n_timesteps|eval_timesteps|eval_episodes|
|----|-----------------------|-------------------:|--------------------:|-----------|-------------:|------------:|
|dqn |AsteroidsNoFrameskip-v4| 730.594059405940600|260.41026953551744000|10M        |        401897|          101|
|dqn |BreakoutNoFrameskip-v4 | 370.558823529411770| 47.82578136655792600|10M        |        354774|           34|
|dqn |EnduroNoFrameskip-v4   | 875.111111111111100|233.08007419694700000|10M        |        399369|            9|
|dqn |PongNoFrameskip-v4     |  20.517241379310345|  0.83545113376453220|10M        |        396917|           58|
|ppo |AsteroidsNoFrameskip-v4|2245.000000000000000|792.29492614808530000|10M        |        401260|           96|
|ppo |BreakoutNoFrameskip-v4 | 389.926829268292700| 35.19311895319563000|10M        |        397003|           41|
|ppo |EnduroNoFrameskip-v4   | 886.125000000000000|142.27388859168784000|10M        |        359432|            8|
|ppo |PongNoFrameskip-v4     |  20.950000000000000|  0.21794494717703367|10M        |        400455|           60|

### Tinyscaler Results

Bilinear filtering:

|algo|        env_id         |     mean_reward     |     std_reward     |n_timesteps|eval_timesteps|eval_episodes|
|----|-----------------------|--------------------:|-------------------:|-----------|-------------:|------------:|
|dqn |AsteroidsNoFrameskip-v4| 707.5757575757576000|239.5791381834171700|10M        |        403627|           99|
|dqn |BreakoutNoFrameskip-v4 |   2.2194719471947195|  3.0834943935340010|10M        |        441820|          606|
|dqn |EnduroNoFrameskip-v4   | 831.8888888888889000|125.9068830917378400|10M        |        399369|            9|
|dqn |PongNoFrameskip-v4     |  20.2105263157894730|  0.9126180213080669|10M        |        396685|           57|
|ppo |AsteroidsNoFrameskip-v4|2128.9361702127660000|684.9110333359602000|10M        |        399180|           94|
|ppo |BreakoutNoFrameskip-v4 | 228.8524590163934500|134.5586393063857700|10M        |        401852|           61|
|ppo |EnduroNoFrameskip-v4   | 754.5000000000000000| 86.8587934523615100|10M        |        386058|           10|
|ppo |PongNoFrameskip-v4     |  19.7192982456140340|  2.1985193260099005|10M        |        401025|           57|

Area filtering variant 1:

|algo|        env_id         |    mean_reward    |     std_reward     |n_timesteps|eval_timesteps|eval_episodes|
|----|-----------------------|------------------:|-------------------:|-----------|-------------:|------------:|
|dqn |AsteroidsNoFrameskip-v4| 967.41176470588230|300.7486851613778600|10M        |        403081|           85|
|dqn |BreakoutNoFrameskip-v4 | 329.69444444444446| 81.9555500949022000|10M        |        322263|           36|
|dqn |EnduroNoFrameskip-v4   | 834.77777777777780|205.3494462172430700|10M        |        399369|            9|
|dqn |PongNoFrameskip-v4     |  20.53448275862069|  0.7925410299697382|10M        |        395163|           58|
|ppo |AsteroidsNoFrameskip-v4|2230.91836734693900|777.1416580015615000|10M        |        405618|           98|
|ppo |BreakoutNoFrameskip-v4 | 369.00000000000000|104.3579576905054900|10M        |        400382|           48|
|ppo |EnduroNoFrameskip-v4   | 951.87500000000000|144.6776049532200500|10M        |        386056|            8|
|ppo |PongNoFrameskip-v4     |  20.71186440677966|  0.5224079663533030|10M        |        398039|           59|

Area filtering variant 2:

|algo|        env_id         |    mean_reward     |     std_reward     |n_timesteps|eval_timesteps|eval_episodes|
|----|-----------------------|-------------------:|-------------------:|-----------|-------------:|------------:|
|dqn |AsteroidsNoFrameskip-v4| 814.945054945055000|235.4811470318050000|10M        |        402045|           91|
|dqn |BreakoutNoFrameskip-v4 | 100.280991735537190|114.2910756326336800|10M        |        405741|          121|
|dqn |EnduroNoFrameskip-v4   | 817.666666666666600|164.2917459210224000|10M        |        386057|            9|
|dqn |PongNoFrameskip-v4     |  20.210526315789473|  0.9865204154041789|10M        |        395431|           57|
|ppo |AsteroidsNoFrameskip-v4|2109.285714285714000|695.7146523906338000|10M        |        404600|           98|
|ppo |BreakoutNoFrameskip-v4 | 367.152173913043500| 70.3234661160728800|10M        |        402943|           46|
|ppo |EnduroNoFrameskip-v4   | 875.555555555555500|180.8880016358256700|10M        |        399369|            9|
|ppo |PongNoFrameskip-v4     |  20.813559322033900|  0.4307886449877571|10M        |        396183|           59|

Area filtering variant 3 (1.1.0 on GitHub):

|algo|        env_id         |    mean_reward     |     std_reward     |n_timesteps|eval_timesteps|eval_episodes|
|----|-----------------------|-------------------:|-------------------:|-----------|-------------:|------------:|
|dqn |AsteroidsNoFrameskip-v4| 759.021739130434800|260.9766763800296300|10M        |        403154|           92|
|dqn |BreakoutNoFrameskip-v4 |  18.321212121212120| 14.9759353949930140|10M        |        411282|          165|
|dqn |EnduroNoFrameskip-v4   | 721.900000000000000|209.3621025878370300|10M        |        399370|           10|
|dqn |PongNoFrameskip-v4     |  20.293103448275860|  0.7429814284970773|10M        |        395853|           58|
|ppo |AsteroidsNoFrameskip-v4|2236.875000000000000|680.6144780331472000|10M        |        405378|           96|
|ppo |BreakoutNoFrameskip-v4 | 329.358490566037740| 96.5397746438619300|10M        |        403565|           53|
|ppo |EnduroNoFrameskip-v4   |1060.714285714285800| 18.0136003041692730|10M        |        372743|            7|
|ppo |PongNoFrameskip-v4     |  20.864406779661017|  0.3423560996054941|10M        |        395353|           59|

**Comment by pseudo-rnd-thoughts** [comment]
In running experiments, we have found that for breakout in particular this new tinyscaler implementation does not achieve the same performance as opencv for dqn, even when training from scratch. See table below. The reason for this is unclear, therefore, we are closing this PR until these issues have been addressed.

### DQN 
Eval num_timesteps=10000000, episode_reward=300.80 +/- 96.14
Episode length: 6748.20 +/- 653.35

| eval  |          |
| --- | --- |
|    mean_ep_length   | 6.75e+03 |
|    mean_reward      | 301      |

| rollout           |          |
| --- | --- | 
|    exploration_rate | 0.01     |

| time               |          |
| --- | --- |
|    total_timesteps  | 10000000 |

| train              |          |
| --- | --- |
|    learning_rate    | 0.0001   |
|    loss             | 0.0102   |
|    n_updates        | 2474999  |

This results are significantly less than the excepted rewards of 375 as reported in [Human-level control through deep reinforcement learning]https://www.nature.com/articles/nature14236

### PPO
| rollout           |          |
| --- | --- |
|    ep_len_mean     | 9.95e+03 |
|    ep_rew_mean     | 375      |

| time              |          |
| --- | --- |
|    fps             | 641      |
|    iterations      | 9766     |
|    time_elapsed    | 15578    |
|    total_timesteps | 10000384 |
